feat: Add option to skip version bumping.

tomfrenken opened the pull request:
Hey all,

in our pipeline we already bumped the version of our package.json in a previous step, therefore we would like to skip bumping the version in this action.

This PR should introduce a simple option to do so.

Commit:
add option to skip bumping

Commit:
add docs

tomfrenken commented:
Hey @TriPSs, do you think we could add this option? :)

TriPSs commented:
How different would this be than setting version file to an empty string?

Commit:
slightly change versioning

tomfrenken commented:
Well, with an empty string we would cause the following error:
```js
throw new Error(`File extension "${ext}" from file "${file}" is not supported`)
```
And even if we use `skip-version-file` and instead set a fallback-version to the one we actually want to use, the version is still bumped as is.

```js
if (skipVersionFile || skipCommit) {
      core.info('Using GIT to determine the new version')
      const versioning = await handleVersioningByExtension( // this function bumps the version currently
        'git',
        versionFile,
        versionPath,
        recommendation.releaseType,
      )
```

tomfrenken commented:
I just added another conditional to set the `newVersion` to the `oldVersion`, as the `newVersion` property was only set when you are bumping (which we are trying not to do), and then later on this property is used for the changelog generation and therefore necessary.

TriPSs commented on src/index.js:
```diff
@@ -120,11 +127,12 @@ async function run() {
       oldVersion = versioning.oldVersion
+      newVersion = skipBump ? oldVersion : versioning.newVersion
 
     } else {
```
```suggestion
    } else if (!skipBump) {
```

TriPSs replied:
Think this is cleaner than passing it down to an other function that than also does not do anything.

TriPSs replied:
Ignore this, you need it for the old version, sorry.

TriPSs reviewed (changes requested):
Although I find it a bit weird I do not see any harm in support things, could you do the following:
1. Commit to this repo following conventional commits
2. Process comments
3. Add tests

Commit:
add parameter to action.yml

Commit:
add test

tomfrenken commented:
Thanks for the review, it wasn't entirely clear to me how all the testing in this repo behaves, but I believe my implementation should test it properly according to the existing standards.

Regarding your first point:  
> Commit to this repo following conventional commits

I did create the PR with a conventional commit message and would usually squash merge & remove all commit messages while squashing, would that work for you, or how exactly should I proceed?

tomfrenken commented:
Also please approve the workflow run again :)

TriPSs commented:
> I did create the PR with a conventional commit message and would usually squash merge & remove all commit messages while squashing, would that work for you, or how exactly should I proceed?

We could do that, regarding the test: should the tag it creates not be the same as the version in the file?

TriPSs commented:
The test is also failing, it's creating a tag with `vnull` now

tomfrenken commented:
Indeed, both the test and the implementation need some changes, currently the old version is only loaded after using the bump function, I'll make some more changes.

Commit:
introduce proper classes, update test

Commit:
update error message

tomfrenken commented:
Alright, I've adjusted the test as per your suggestion and also changed the classes to use a proper constructor instead of the `init` method, now the files are read and parsed at initialization, this way the `oldVersion` is accessible independently of bumping.

Commit:
build

Commit:
adjust yaml bump function

Commit:
build

Commit:
manually trigger actions

Commit:
fix few bugs

Commit:
chore(release): v6.0.0 [skip ci]

Commit:
Revert "chore(release): v6.0.0 [skip ci]"

This reverts commit a309700677c7560718b66f6f3026eab9bbcbea96.

TriPSs commented:
@tomfrenken you can also run the actions locally, see the [README](https://github.com/TriPSs/conventional-changelog-action?tab=readme-ov-file#development). My preference would be to only implement your feature here and leave existing code more as is (So keep the init, as I prefer to only pass args once and not for each class separate).

Commit:
fix toml and yaml

Commit:
fix toml and yaml

Commit:
fix toml and yaml

Commit:
final fix?

Commit:
little test

Commit:
yaml fix nr 25

Commit:
yaml fix nr 26

Commit:
remove constructor, reintroduce adjusted init

Commit:
remove workflow_dispatch

tomfrenken commented:
Alright, thanks for the hint, however, the logs (when running the tests locally in parallel) are somewhat annoying to read 😬.

I've run the action now on my fork, all tests are passing.
@TriPSs as per your suggestion I removed the constructor again and added the init back, however with a slight adjustment as the init looks slightly different for the different file types.

Commit:
build

TriPSs reviewed (commented):
We could also make a `readFile` method in the base that is abstract, and call it inside the `init` of `BaseVersioning`, that way we do not need to define an init in each version.

What you think?

Commit:
introduce different fallback version logic for skip bump

Commit:
temporarly add workflow dispatch

Commit:
fix skip bump logic again

Commit:
test run

Commit:
change

Commit:
add another test

Commit:
introduce more tests to cover corner cases

Commit:
final test?

Commit:
final test?

Commit:
fix parsing issue

Commit:
change test to match test case

Commit:
try bash test

Commit:
should fail

Commit:
cleanup

tomfrenken commented:
@TriPSs thanks for the abstract method suggestion, great idea.
I've implemented the change and also added additional tests for corner cases.

Commit:
build

Commit:
cleanup

TriPSs commented on src/index.js:
```diff
@@ -120,11 +124,17 @@ async function run() {
       oldVersion = versioning.oldVersion
+      // If we are skipping the bump, we either use the fallback version or the old version as the new version.
+      if(skipBump){
+        newVersion = fallbackVersion && !oldVersion ? fallbackVersion : oldVersion
```
```suggestion
        newVersion = oldVersion || fallbackVersion 
```

TriPSs replied:
We could also do
```suggestion
        newVersion = skipBump
                ? oldVersion || fallbackVersion 
                : versioning.newVersion
```

TriPSs commented on src/index.js:
```diff
@@ -135,12 +145,16 @@ async function run() {
-
-      newVersion = versioning[0].newVersion
       oldVersion = versioning[0].oldVersion
+      // If we are skipping the bump, we either use the fallback version or the old version as the new version.
+      if(skipBump){
+        newVersion = fallbackVersion && !oldVersion ? fallbackVersion : oldVersion
+      } else {
+        newVersion = versioning[0].newVersion
+      }
```
Is this needed as `handleVersioningByExtension` makes sure the correct version is set in `newVersion`?

tomfrenken replied:
Yeah, that's a good point, I think I should move the entire versioning logic into that function instead of putting it here

TriPSs commented on src/version/json.js:
```diff
@@ -6,33 +6,38 @@ const bumpVersion = require('../helpers/bumpVersion')
 
 module.exports = class Json extends BaseVersioning {
 
+  eol = null;
+  jsonContent = {};
```
```suggestion
  eol = null
  jsonContent = {}
```

TriPSs reviewed (commented):
Thanks already for your work!

Commit:
fix logic

Commit:
suggestion

Commit:
move fallback getter

Commit:
build & cleanup

Commit:
clarify comment

tomfrenken commented:
Awesome, moving the logic to the version function also removed the code duplication, and the improved conditional logic is also way cleaner.

TriPSs commented:
Thanks for the PR 🎉

tomfrenken commented:
Thanks for the quick reviews!
